cluster-codec-parent/cluster-codec-jackson/src/main/java/io/scalecube/cluster/codec/jackson/JacksonMetadataEncoder.java

```diff
@@ -0,0 +1,72 @@
+package io.scalecube.cluster.codec.jackson;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.scalecube.cluster.metadata.MetadataDecoder;
+import io.scalecube.cluster.metadata.MetadataEncoder;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.Map;
+import reactor.core.Exceptions;
+
+/**
+ * Jackson-based metadata encoder/decoder.
+ *
+ * @author eutkin
+ */
+public class JacksonMetadataEncoder implements MetadataEncoder, MetadataDecoder {
+
+  private static final TypeReference<Map<String, String>> TYPE =
+      new TypeReference<Map<String, String>>() {};
+
+  private final ObjectMapper delegate;
+
+  /**
+   * Create instance with external {@link ObjectMapper}.
+   *
+   * @param delegate jackson object mapper
+   */
+  public JacksonMetadataEncoder(ObjectMapper delegate) {
+    this.delegate = delegate;
+  }
+
+  /**
+   * Create instance with default {@link ObjectMapper}.
+   */
+  public JacksonMetadataEncoder() {
+    this.delegate = DefaultObjectMapper.OBJECT_MAPPER;
+  }
+
+  /**
+   * Deserialize metadata as {@link Map Map&lt;String, String&gt;}.
+   *
+   * @param buffer binary metadata
+   * @return medata as map
+   */
+  @Override
+  public Object decode(ByteBuffer buffer) {
+    try {
+      if (buffer.remaining() == 0) {
+        return Collections.emptyMap();
+      }
+      return this.delegate.readValue(buffer.array(), TYPE);
```

Comment:
Now it's hardcoded type `Map<String, String>`, what about other custom types of metadata, can it be more generic? It's kinda passing a needed type and using it into mapper:

```java
return mapper.readValue(buffer.array(), mapper.getTypeFactory().constructType(type));
```

Reply:
@segabriel, i commited, please see

Reply:
Hm, probably let's revert your last commit.
Maybe we don't need to break down the current API and just enable putting type identifier into JSON as a property (`"@class": "java.util.HashMap"`). I mean something like that:

```
mapper.enableDefaultTyping(DefaultTyping.NON_FINAL, As.PROPERTY);
```

and then in the decode phase you need:

```
Object value = mapper.readValue(json.getBytes(), Object.class)
```

PS: it's our default impl so if they want to improve such mechanism they should set it explicitly.